modules/gapi/src/streaming/vpl/surface/surface.cpp

```diff
@@ -0,0 +1,74 @@
+#include "streaming/vpl/surface/surface.hpp"
+#include "logger.hpp"
+
+#ifdef HAVE_ONEVPL
+
+namespace cv {
+namespace gapi {
+namespace wip {
+
+Surface::Surface(std::unique_ptr<handle_t>&& surf, std::shared_ptr<void> associated_memory) :
+    workspace_memory_ptr(associated_memory),
+    mfx_surface(std::move(surf)),
+    mirrored_locked_count() {
+
+    GAPI_Assert(mfx_surface && "Surface is nullptr");
+    mirrored_locked_count.store(mfx_surface->Data.Locked);
+    GAPI_LOG_DEBUG(nullptr, "create surface: " << mfx_surface <<
+                            ", locked count: " << mfx_surface->Data.Locked);
+}
+
+Surface::~Surface() {
+    GAPI_LOG_DEBUG(nullptr, "destroy surface: " << mfx_surface <<
+                            ", worspace memory counter: " << workspace_memory_ptr.use_count());
+}
+
+std::shared_ptr<Surface> Surface::create_surface(std::unique_ptr<handle_t>&& surf,
+                                                 std::shared_ptr<void> accociated_memory) {
+    surface_ptr_t ret {new Surface(std::move(surf), accociated_memory)};
+    return ret;
+}
+
+std::shared_ptr<Surface> Surface::get_ptr() {
+    return shared_from_this();
+}
+
+Surface::handle_t* Surface::get_handle() const {
+    return mfx_surface.get();
+}
+
+const Surface::info_t& Surface::get_info() const {
+    return mfx_surface->Info;
+}
+
+const Surface::data_t& Surface::get_data() const {
+    return mfx_surface->Data;
+}
+
+size_t Surface::get_locks_count() const {
+    return mirrored_locked_count.load();
+}
+
+size_t Surface::obtain_lock() {
+    size_t locked_count = mirrored_locked_count.fetch_add(1);
+    GAPI_Assert(locked_count < std::numeric_limits<mfxU16>::max() && "Too many references ");
+    mfx_surface->Data.Locked = static_cast<mfxU16>(locked_count + 1);
+    GAPI_LOG_DEBUG(nullptr, "surface: " << mfx_surface.get() <<
+                            ", locked times: " << locked_count + 1);
+    return locked_count + 1;
+}
+
+size_t Surface::release_lock() {
+    size_t locked_count = mirrored_locked_count.fetch_sub(1);
+    GAPI_Assert(locked_count < std::numeric_limits<mfxU16>::max() && "Too many references ");
+    GAPI_Assert(locked_count && "Surface lock counter is invalid");
+    mfx_surface->Data.Locked = static_cast<mfxU16>(locked_count - 1);
+    GAPI_LOG_DEBUG(nullptr, "surface: " << mfx_surface.get() <<
+                            ", locked times: " << locked_count - 1);
+    return locked_count + 1;
+}
```

Comment:
Do you have any concurrent tests?

Reply:
no, but concurrency model is very simple here. Please see comment about  get_locks_count()

There are multithreading scenarios in samples and in UT: source kills first before last MediaFrame consumed and all MediaFrames killed fist before source expired. Looks processed